### PR TITLE
V1.0 backports 18 06 06

### PR DESCRIPTION
**Pull request by jrfastab**
**Summary of changes**:

**Backports targeting branch v1.0 for eventual release v1.4.**
PR: #4340 -- Add k8s label source in GetPolicyLabels
PR: #4306 -- identity: Resolve unknown identity to label reserved:unknown
PR: #4294 -- identity: Ignore nil identity when generating IdentityCache
PR: #4250 -- Add "docker info" output to bugtool
PR: #4247 -- init.sh: Use 'ip route replace' instead of 'ip route add'
PR: #4239 -- policy: always regenerate if policy forcibly computed
PR: #4239 -- policy: always regenerate if policy forcibly computed
PR: #4230 -- cilium status: fix --brief to print less when cilium isn't running
PR: #4218 -- test: download exact k8s version of k8s upstream e2e
PR: #4214 -- bpf: Only create veth pair if it does not already exist.
PR: #4205 -- test: update k8s versions to 1.7.15, 1.8.13, 1.9.8, 1.10.3 and 1.11.0-beta.0

**PRs that were not backported despite tags in 'PR' list. I believe these are needed in v1.1 _not_ v1.0:**
PR: #4365 -- daemon: Delete old ID mapping when updating the IP for a reserved ID
PR: #4167 -- ipcache: Create copies of NPHDS cache resources when updating
PR: #3968 -- docs: Update Istio GSG to Istio 0.8.0 pre-release

**PR 4240 not backported because it fixes BPF program regeneration optimization which was not in v1.0 to start with. If needed later open up an issue to add feature to v1.0 we don't want to simply cherrypick all dependencies here, there are too many.***
PR: #4240 -- pkg/endpoint: include node headerfile hash

**This PR was reverted and we are waiting for the revised PR to be accepted:**
PR: #3786 -- pkg/bpf: Use the other directory when /sys/fs/bpf is not BPFFS

**Notes**
The following commit was also pulled in to resolve `PR4240` dependencies.
```
commit f7e4adf68ec0718962a8f4c3fb0782fe2c739ff3
Author: Romain Lenglet <romain@covalent.io>
Date:   Mon Apr 16 11:26:28 2018 -0700

    endpoint: Skip BPF compilation if headerfile is unchanged
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4410)

### CODE REVIEW

**Comment by jrfastab** [comment]
@ianvernon: Please review carefully `PR: 4240 -- pkg/endpoint: include node headerfile hash` required numerous merge conflict resolution. Note I also pulled in this dependency, 
```
commit f7e4adf68ec0718962a8f4c3fb0782fe2c739ff3
Author: Romain Lenglet <romain@covalent.io>
Date:   Mon Apr 16 11:26:28 2018 -0700

    endpoint: Skip BPF compilation if headerfile is unchanged
```

**Comment by jrfastab** [comment]
@aanm with regards to commit,

```
commit 7baa81ab1aedd1c426e6b3b4ee8d0f78850f6c70
Author: André Martins <andre@cilium.io>
Date:   Wed May 23 12:39:16 2018 +0200

    test: download exact k8s version of k8s upstream e2e
```
the file `./test/kubernetes-test.sh` was not part of the v1.0 branch. I've added it here, let me know if its more appropriate to drop the file.

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

File: pkg/endpoint/bpf.go [hunk 1]
```diff
@@ -337,31 +338,62 @@ func (e *Endpoint) writeHeaderfile(prefix string, owner Owner) error {
 	return fw.Flush()
 }
 
-// hashHeaderfile returns the MD5 hash of the BPF headerfile with the given prefix.
+// hashEndpointHeaderFiles returns the MD5 hash of any header files that are
+// used in the compilation of an endpoint's BPF program. Currently, this
+// includes the endpoint's headerfile, and the node's headerfile.
+func hashEndpointHeaderfiles(prefix string) (string, error) {
+	endpointHeaderPath := filepath.Join(prefix, common.CHeaderFileName)
+	nodeConfigPath := "./globals/node_config.h"
```
Comment (ianvernon) [inline pkg/endpoint/bpf.go]: I just realized that this has to be changed to be an absolute path - you will need to pass `Owner` to this function and then call `owner.GetStateDir()` to get `nodeConfigPath` to be at the correct path. Sorry about that :(

Comment (ianvernon) [reply pkg/endpoint/bpf.go]: Actually, if https://github.com/cilium/cilium/pull/4410/commits/f7e4adf68ec0718962a8f4c3fb0782fe2c739ff3 was a backport as well to 1.0 (i.e., we are not hashing the headerfiles in 1.0), then we do not need to backport this commit either to 1.0, only 1.1. I believe you can revert these two commits. The hashing of the headerfile was an optimization, and this commit  ( https://github.com/cilium/cilium/pull/4410/commits/58a38a5523df35a8aa09a72a6de2350b07cffa0d ) is a fix for that commit which you also had to backport. cc: @rlenglet to confirm I am correct. Sorry about all of the confusion.

**Review by ianvernon — CHANGES_REQUESTED** [review]
Please see my comments inline about backporting the headerfile changes. I think the fix I made needs to be backported to the 1.1 branch, not 1.0.

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]
Thanks for fixing so quickly, and my apologies again for leading you astray with the headerfile commits!

**Comment by jrfastab** [comment]
@ianvernon hmm got a test failure (or maybe just a flake) with the changes, I'm going to try and get a good build with those two popped off first. May add them back after I get a good build.

**Comment by jrfastab** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Comment by jrfastab** [comment]
test-me-please

**Pull request merged**